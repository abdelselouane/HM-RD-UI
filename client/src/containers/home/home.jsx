/* global window */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { withContext } from '../../contexts/history';

import Page, { Container } from '../../components/page';
import TileContainer from '../../components/tileContainer';

export class Home extends PureComponent {
  static propTypes = {
    config: PropTypes.shape({
      partsFinderUrl: PropTypes.string.isRequired
    }).isRequired,
    history: PropTypes.shape({
      push: PropTypes.func.isRequired
    }).isRequired
  }

  partsFinderTileFunction = () => {
    window.open(this.props.config.partsFinderUrl, '_blank');
  }

  orderPartsTileFunction = () => {
    this.props.history.push('/orderParts');
  }

  orderHistoryTileFunction = () => {
    this.props.history.push('/orderHistory');
  }

  inventoryFunction = () => {
    this.props.history.push('/inventory');
  }

  workOrdersTileFunction = () => {
    this.props.history.push('/workOrders');
  }

  tiles = [
    {
      title: 'Parts Finder',
      className: 'parts-search',
      icon: 'search',
      tileFunction: this.partsFinderTileFunction
    },
    {
      title: 'Work Orders',
      className: 'work-orders',
      icon: 'toolbox',
      tileFunction: this.workOrdersTileFunction
    },
    {
      title: 'Order Parts',
      className: 'order-parts',
      icon: 'tool-wrench',
      tileFunction: this.orderPartsTileFunction
    },
    {
      title: 'Order History',
      className: 'order-history',
      icon: 'assignment',
      tileFunction: this.orderHistoryTileFunction
    },
    {
      title: 'Inventory',
      className: 'inventory',
      icon: 'view-list',
      tileFunction: this.inventoryFunction
    }
  ];

  render() {
    return (
      <Page className="home">
        <Container hideBoxShadow>
          <TileContainer tiles={this.tiles} />
        </Container>
      </Page>
    );
  }
}

const connectedHome = connect(
  state => ({ config: state.config }),
  null
)(Home);

export default withContext(connectedHome);
