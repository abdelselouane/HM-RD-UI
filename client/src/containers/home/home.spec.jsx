/* global describe, beforeAll, jest, it, expect */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Home } from './home';

import Page from '../../components/page';

describe('<Home />', () => {
  let spy;
  let wrapper;
  let wrapperInstance;
  let history;
  let props;

  beforeAll(() => {
    history = { push: () => {}, location: { pathname: '' } };
    props = {
      config: {
        partsFinderUrl: 'mockUrl'
      },
      history
    };

    wrapper = shallow(<Home {...props} />);
    wrapperInstance = wrapper.instance();
  });

  it('should render properly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render <Page/> component', () => {
    expect(wrapper.find(Page)).toHaveLength(1);
  });

  it('should push workOrders to history on Work Orders tileFunction call', () => {
    spy = jest.spyOn(history, 'push');
    wrapperInstance.tiles[1].tileFunction();
    expect(spy).toHaveBeenCalledWith('/workOrders');
    spy.mockReset();
    spy.mockRestore();
  });

  it('should push orderParts to history on Order Parts tileFunction call', () => {
    spy = jest.spyOn(history, 'push');
    wrapperInstance.tiles[2].tileFunction();
    expect(spy).toHaveBeenCalledWith('/orderParts');
    spy.mockReset();
    spy.mockRestore();
  });

  it('should push orderHistory to history on Order History tileFunction call', () => {
    spy = jest.spyOn(history, 'push');
    wrapperInstance.tiles[3].tileFunction();
    expect(spy).toHaveBeenCalledWith('/orderHistory');
    spy.mockReset();
    spy.mockRestore();
  });

  it('should push inventory to history on Inventory tileFunction call', () => {
    spy = jest.spyOn(history, 'push');
    wrapperInstance.tiles[4].tileFunction();
    expect(spy).toHaveBeenCalledWith('/inventory');
    spy.mockReset();
    spy.mockRestore();
  });

  it('should call open new tab with partsFinderUrl on Parts Finder tileFunction call', () => {
    global.open = jest.fn();
    wrapperInstance.tiles[0].tileFunction();
    expect(global.open).toHaveBeenCalledWith(props.config.partsFinderUrl, '_blank');
  });
});
