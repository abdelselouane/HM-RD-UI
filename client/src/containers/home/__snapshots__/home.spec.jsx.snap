// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render properly 1`] = `
<Connect(withContext(Page))
  className="home"
>
  <styled.div
    hideBoxShadow={true}
  >
    <TileContainer
      tiles={
        Array [
          Object {
            "className": "parts-search",
            "icon": "search",
            "tileFunction": [Function],
            "title": "Parts Finder",
          },
          Object {
            "className": "work-orders",
            "icon": "toolbox",
            "tileFunction": [Function],
            "title": "Work Orders",
          },
          Object {
            "className": "order-parts",
            "icon": "tool-wrench",
            "tileFunction": [Function],
            "title": "Order Parts",
          },
          Object {
            "className": "order-history",
            "icon": "assignment",
            "tileFunction": [Function],
            "title": "Order History",
          },
          Object {
            "className": "inventory",
            "icon": "view-list",
            "tileFunction": [Function],
            "title": "Inventory",
          },
        ]
      }
    />
  </styled.div>
</Connect(withContext(Page))>
`;
